zebra-state/src/service/write.rs

```diff
@@ -328,13 +328,37 @@ fn log_if_mined_by_zebra(tip_block: &ChainTipBlock) {
         .starts_with(EXTRA_ZEBRA_COINBASE_DATA.as_bytes())
     {
         let text = String::from_utf8_lossy(coinbase_data.as_ref());
-        let data = hex::encode(coinbase_data.as_ref());
-        info!(
-            ?text,
-            ?data,
-            %height,
-            %hash,
-            "looks like this block was mined by Zebra!"
-        );
+
+        if coinbase_data.as_ref() == EXTRA_ZEBRA_COINBASE_DATA.as_bytes() {
+            info!(
+                %text,
+                %height,
+                %hash,
+                "looks like this block was mined by Zebra!"
+            );
+        } else {
+            // # Security
+            //
+            // Use the extra data as an allow-list, replacing unknown characters.
+            // This makes sure control characters and harmful messages don't get logged
+            // to the terminal.
+            let text = text.replace(
+                |c| {
+                    !EXTRA_ZEBRA_COINBASE_DATA
```

Comment:
`from_utf8_lossy()` deals with this already: partial unicode characters are illegal in Rust strings.
https://doc.rust-lang.org/std/string/struct.String.html#method.from_utf8_lossy

Reply:
I think there could be an issue around comparing the character in `text` to the character code. Would filtering characters after the expected🦓 with [`is_control`](https://doc.rust-lang.org/stable/std/primitive.char.html#method.is_control) work if that's an issue?

Reply:
I'm not sure how that could happen, can you give an example?

Here is my analysis:

`replace()` is used as an allow-list, replacing every character except for `Z`, `z`, and the Zebra emoji with `?`.

These characters and their UTF-8 bytes are ok to write to a terminal, because they are all printable:
> [240, 159, 166, 147]
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7f5a7d8711a765c74414310e01a10b5f
https://www.ascii-code.com/

Here are the docs for replace():

replace:
> Replaces all matches of a pattern with another string.

https://doc.rust-lang.org/std/string/struct.String.html#method.replace

Pattern:
> F: FnMut(char) -> bool
> F returns true for a char in string

https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html

Reply:
> I'm not sure how that could happen, can you give an example?

It can't happen, your code is correct, I misunderstood `EXTRA_ZEBRA_COINBASE_DATA` to be a hex value in string format.

Reply:
Thanks for checking, it's important that we make sure our code is secure!